### PR TITLE
Vietnamese keyboard

### PR DESCRIPTION
**Pull request by K4zoku**
Added missing accent (horn, hook above, dot below)
Added layout (~~I am just putting new accents in random position~~ Now it's telex layout)
Added accent font svg (by modified existing accents)

Related issue(s): #290

### CODE REVIEW

File: res/xml/qwerty_vi.xml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  </row>
+  <row>
+    <key shift="0.5" key0="a" key1="tab" key2="`" key3="â"/>
+    <key key0="s" key1="loc accent_aigu"/>
+    <key key0="d" key1="loc accent_bar" key2="₫" key3="đ"/>
+    <key key0="f" key1="loc accent_grave"/>
+    <key key0="g" key2="-" key3="_"/>
+    <key key0="h" key2="=" key3="+"/>
+    <key key0="j" key1="loc accent_dot_below" key3="{" key4="}"/>
+    <key key0="k" key1="loc accent_hook" key3="[" key4="]"/>
```
Comment (Julow) [inline res/xml/qwerty_vi.xml:23]: There's no `accent_hook`.

File: res/xml/qwerty_vi.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<keyboard>
+  <row>
+    <key key0="q" key1="~" key2="1" key4="esc"/>
+    <key key0="w" key1="loc accent_breve" key2="2" key3="\@" key4="loc accent_horn"/>
+    <key key0="e" key1="!" key2="3" key3="\#" key4="ê"/>
+    <key key0="r" key2="4" key3="$" key1="loc accent_hook_above"/>
```
Comment (Julow) [inline res/xml/qwerty_vi.xml:7]: All the accents that you need shouldn't have the `loc ` prefix. Like this, someone who has not the languages enabled on their device but decide to use this layout won't see the accents.

File: srcs/juloo.keyboard2/KeyModifier.java [hunk 3]
```diff
@@ -60,6 +60,9 @@ public static KeyValue modify(KeyValue k, KeyValue.Modifier mod)
       case SLASH: return apply_map_char(k, map_char_slash);
       case BAR: return apply_map_char(k, map_char_bar);
       case ARROW_RIGHT: return apply_combining(k, "\u20D7");
+      case DOT_BELOW: return apply_dead_char(k, '\u0323');
+      case HORN: return apply_map_char(k, map_char_horn);
+      case HOOK_ABOVE: return apply_dead_char(k, '\u0309');
```
Comment (Julow) [inline srcs/juloo.keyboard2/KeyModifier.java:65]: `DOT_BELOW` and `HOOK_ABOVE` don't work.
It seems that these characters use combining diacritics ? In this case, you should use `apply_combining`.

**Review by Julow — COMMENTED** [review]
Thanks for the great work!

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: no, I think I need to write a map char for this. Currently, I can type all vietnamese characters, but a little bit janky (e.g: ạ vs ạ is different)

File: res/xml/qwerty_vi.xml (hunk 2 above)
Comment (K4zoku) [reply res/xml/qwerty_vi.xml]: removed in latest commit

File: res/xml/qwerty_vi.xml (hunk 1 above)
Comment (K4zoku) [reply res/xml/qwerty_vi.xml]: i forgor 💀
btw, removed in latest commit

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: I'll add map char for those two accent soon

Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: If use `apply_combining`, the other accents should be use it too and the whole code is only for Vietnamese, which is a big deal.

File: res/xml/qwerty_vi.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<keyboard>
+  <row>
+    <key key0="q" key1="~" key2="1" key4="esc"/>
+    <key key0="w" key1="loc accent_breve" key2="2" key3="\@" key4="accent_horn"/>
```
Comment (Julow) [inline res/xml/qwerty_vi.xml:5]: Breve is needed ? It shouldn't have the `loc` prefix then. The `loc` prefix is used to place a key but to show it only if it is an "extra key" or is enabled in the settings (eg. like `capslock`)

Comment (K4zoku) [reply res/xml/qwerty_vi.xml:5]: oh, I forgor again 💀
thanks for figure it out
(it's 4:53 AM now and I haven't sleep yet)

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (Julow) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: Do you mean that Vietnamese uses combining diacritics for breve, aigu, grave, etc.. ?

**Comment by K4zoku** [comment]
btw, the newly added svgs need to be reposition to get aligned with the others

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: yes
ex: `a á à ả ã ạ`

File: res/xml/qwerty_vi.xml (hunk 4 above)
Comment (K4zoku) [reply res/xml/qwerty_vi.xml]: Removed in latest commit

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (Julow) [reply srcs/juloo.keyboard2/KeyModifier.java:65]: In this case, you can simply use `apply_combining` to implement `DOT_BELOW` and `HOOK_ABOVE`.

For the other accents, you'll need to define new versions of them (eg. `AIGU_COMBINING`, etc..) that use the same symbol but are implemented using `apply_combining`.

Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java]: The combining is janky, I added map char and it now work as I (a Vietnamese guy) expected

Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java]: if combining, the accent and character before are two characters, meanwhile almost Vietnamese typing engine didn't use that, they use an actual character

Comment (K4zoku) [reply srcs/juloo.keyboard2/KeyModifier.java]: The keyboard is perfect for me now. If someone has problem about the Vietnamese layout and accent when using UnexpectedKeyboard, just mention me

**Comment by K4zoku** [comment]
Screenshot_20230209-235331

Screenshot for the latest layout change

File: srcs/juloo.keyboard2/KeyModifier.java (hunk 3 above)
Comment (Julow) [reply srcs/juloo.keyboard2/KeyModifier.java]: That works for me.

**Review by Julow — APPROVED** [review]
Thanks a lot for your work!

I pushed a commit to add the missing check boxes. This is a hack that I should remove soon.

**Pull request merged**